File: ros2_ws/src/cpp_tf2/src/dynamic_broadcaster.cpp
#include "cpp_tf2/dynamic_broadcaster.hpp"

#include <chrono>
#include <cmath>

namespace learning_ros2::cpp_tf2 {

DynamicFrameBroadcaster::DynamicFrameBroadcaster(
    const rclcpp::NodeOptions& options)
    : rclcpp::Node("dynamic_frame_broadcaster", options),
      radius_(1.0),
      angular_speed_(1.0),
      angle_(0.0),
      dt_(0.1) {
  tf_broadcaster_ = std::make_shared<tf2_ros::TransformBroadcaster>(this);

  timer_ = this->create_wall_timer(
      std::chrono::duration<double>(dt_),
      std::bind(&DynamicFrameBroadcaster::timer_callback, this));
}

void DynamicFrameBroadcaster::timer_callback() {
  geometry_msgs::msg::TransformStamped t;

  t.header.stamp = this->get_clock()->now();
  t.header.frame_id = "base_link";
  t.child_frame_id = "moving_frame";

  t.transform.translation.x = radius_ * std::cos(angle_);
  t.transform.translation.y = radius_ * std::sin(angle_);
  t.transform.translation.z = 0.0;

  double yaw = angle_ + M_PI / 2.0;
  t.transform.rotation.x = 0.0;
  t.transform.rotation.y = 0.0;
  t.transform.rotation.z = std::sin(yaw / 2.0);
  t.transform.rotation.w = std::cos(yaw / 2.0);

  tf_broadcaster_->sendTransform(t);

  angle_ += angular_speed_ * dt_;
  if (angle_ > 2.0 * M_PI) {
    angle_ -= 2.0 * M_PI;
  }
}

}  // namespace learning_ros2::cpp_tf2

File: ros2_ws/src/py_tf2/py_tf2/dynamic_broadcaster.py
from typing import Optional

import math
import rclpy
from rclpy.node import Node
from geometry_msgs.msg import TransformStamped
from tf2_ros.transform_broadcaster import TransformBroadcaster


class DynamicFrameBroadcaster(Node):
    def __init__(self) -> None:
        super().__init__("dynamic_frame_broadcaster")
        self.tf_broadcaster = TransformBroadcaster(self)

        self.timer = self.create_timer(0.1, self.timer_callback)

        # 圆周运动参数
        self.radius: float = 1.0
        self.angular_speed: float = 1.0
        self.angle: float = 0.0
        self.dt: float = 0.1

    def timer_callback(self) -> None:
        t = TransformStamped()

        t.header.stamp = self.get_clock().now().to_msg()
        t.header.frame_id = "base_link"
        t.child_frame_id = "moving_frame"

        t.transform.translation.x = self.radius * math.cos(self.angle)
        t.transform.translation.y = self.radius * math.sin(self.angle)
        t.transform.translation.z = 0.0

        yaw = self.angle + math.pi / 2.0
        t.transform.rotation.x = 0.0
        t.transform.rotation.y = 0.0
        t.transform.rotation.z = math.sin(yaw / 2.0)
        t.transform.rotation.w = math.cos(yaw / 2.0)

        self.tf_broadcaster.sendTransform(t)

        self.angle += self.angular_speed * self.dt
        if self.angle > 2.0 * math.pi:
            self.angle -= 2.0 * math.pi


def main(args: Optional[list[str]] = None) -> None:
    rclpy.init(args=args)
    node = DynamicFrameBroadcaster()
    rclpy.spin(node)
    node.destroy_node()
    rclpy.shutdown()


if __name__ == "__main__":
    main()
